File: typescript/src/domain/Order.ts
import OrderItem from './OrderItem';
import {OrderStatus} from './OrderStatus';
import OrderApprovalRequest from '../useCase/OrderApprovalRequest';
import ShippedOrdersCannotBeChangedException from '../useCase/ShippedOrdersCannotBeChangedException';
import RejectedOrderCannotBeApprovedException from '../useCase/RejectedOrderCannotBeApprovedException';
import ApprovedOrderCannotBeRejectedException from '../useCase/ApprovedOrderCannotBeRejectedException';

class Order {
  constructor(status: OrderStatus, identifier: number, currency: string = 'EUR') {
    this.status = status;
    this.id = identifier;
    this.items = [];
    this.total = 0;
    this.tax = 0;
    this.currency = currency;
  }

  public isShipped(): boolean {
    return this.status === OrderStatus.SHIPPED;
  }

  public isRejected(): boolean {
    return this.status === OrderStatus.REJECTED;
  }

  public isApproved(): boolean {
    return this.status === OrderStatus.APPROVED;
  }

  public verifyApprovalStatus(request: OrderApprovalRequest): void {
    if (this.isShipped()) {
      throw new ShippedOrdersCannotBeChangedException();
    }

    if (request.isApproved() && this.isRejected()) {
      throw new RejectedOrderCannotBeApprovedException();
    }

    if (!request.isApproved() && this.isApproved()) {
      throw new ApprovedOrderCannotBeRejectedException();
    }
  }

  public addOrderItem(item: OrderItem): void {
    this.items.push(item);
    this.tax += item.getTax();
    this.total += item.getTaxedAmount();
  }

  private total: number;
  private currency: string;
  private items: OrderItem[];
  private tax: number;
  private status: OrderStatus;
  private id: number;

  public getTotal(): number {
    return this.total;
  }

  public setTotal(total: number): void {
    this.total = total;
  }

  public getCurrency(): string {
    return this.currency;
  }


  public getItems(): OrderItem[] {
    return this.items;
  }

  public getTax(): number {
    return this.tax;
  }

  public setTax(tax: number): void {
    this.tax = tax;
  }

  public getStatus(): OrderStatus {
    return this.status;
  }

  public setStatus(status: OrderStatus): void {
    this.status = status;
  }

  public getId(): number {
    return this.id;
  }
}

export default Order;

File: python/src/domain/order.py
from dataclasses import dataclass, field

from .order_item import OrderItem
from .order_status import OrderStatus
from ..use_case.exceptions import ShippedOrdersCannotBeChangedException, ApprovedOrderCannotBeRejectedException, \
    RejectedOrderCannotBeApprovedException


@dataclass
class Order:
    id: int = 0
    total: float = 0
    currency: str = 'EUR'
    items: list[OrderItem] = field(default_factory=list)
    tax: float = 0
    status: OrderStatus = OrderStatus.CREATED

    def __init__(self):
        super().__init__()
        self.items = []
        self.currency = "EUR"
        self.total = 0
        self.tax = 0

    def is_shipped(self):
        return self.status == OrderStatus.SHIPPED

    def is_rejected(self):
        return self.status == OrderStatus.REJECTED

    def is_approved(self):
        return self.status == OrderStatus.APPROVED

    def approve(self, is_request_approved):
        if self.is_shipped():
            raise ShippedOrdersCannotBeChangedException()

        if is_request_approved and self.is_rejected():
            raise RejectedOrderCannotBeApprovedException()

        if not is_request_approved and self.is_approved():
            raise ApprovedOrderCannotBeRejectedException()

        self.status = OrderStatus.APPROVED if is_request_approved else OrderStatus.REJECTED

    def add_item(self, item):
        self.items.append(item)
        self.total += item.taxed_amount
        self.tax += item.tax
